File: src/dataset/Java/bob_exercism-java-blogscot/gemini/GeminiFlash_bob_v2.java
import java.util.regex.Pattern;

public class Bob {

  /**
   * Determines if the given text represents shouting.
   * Shouting is defined as containing at least one letter, and all letters being uppercase.
   * This method performs a single pass over the text, avoiding intermediate string creations.
   *
   * @param text The string to check.
   * @return true if the text is shouting, false otherwise.
   */
  private boolean isShouting(String text) {
    boolean hasLetter = false;
    for (int i = 0; i < text.length(); i++) {
      char c = text.charAt(i);
      if (Character.isLetter(c)) {
        hasLetter = true;
        if (Character.isLowerCase(c)) {
          return false; // Found a lowercase letter, so it's not shouting.
        }
      }
    }
    // Returns true only if at least one letter was found AND all found letters were uppercase.
    return hasLetter;
  }

  /**
   * Bob's response logic based on the input text.
   * This method optimizes the process by avoiding expensive regex operations and
   * minimizing string allocations, while preserving original functionality.
   *
   * @param text The input string for Bob's response.
   * @return Bob's appropriate response.
   */
  String hey(String text) {
    // Build the 'letters' string by filtering out characters not relevant to the rules,
    // which aligns with the original `replaceAll("([^\\w?])", "")`.
    // This avoids the overhead of regex compilation and string creation associated with `replaceAll`.
    StringBuilder sb = new StringBuilder();
    for (int i = 0; i < text.length(); i++) {
      char c = text.charAt(i);
      // `\w` in regex typically means [a-zA-Z0-9_].
      // So, we keep letters, digits, underscores, and question marks.
      if (Character.isLetterOrDigit(c) || c == '_' || c == '?') {
        sb.append(c);
      }
    }
    String letters = sb.toString();

    // Apply the original conditions in the specified order to maintain functionality.
    // The `isShouting` method is already optimized for performance.
    // The `isBlank()` check on 'letters' correctly handles inputs like "   ", ".,!", etc.,
    // which become an empty string after filtering.
    if (isShouting(letters) && letters.endsWith("?")) {
      return "Calm down, I know what I'm doing!";
    }
    if (isShouting(letters)) {
      return "Whoa, chill out!";
    }
    if (letters.endsWith("?")) {
      return "Sure.";
    }
    if (letters.isBlank()) { // Checks if the filtered string is empty or contains only whitespace characters (which it won't after filtering, so effectively checks if it's empty).
      return "Fine. Be that way!";
    }
    return "Whatever.";
  }
}

File: src/dataset/typescript/bob_Exercism-typescript-FilipeCerejo/gemini/GeminiFlash_bob_v2.ts
export function hey(message: string): string {
    const trimmedMessage = message.trim();
    const hasMessage = trimmedMessage !== '';

    if (!hasMessage) {
        return 'Fine. Be that way!';
    }

    const isQuestion = trimmedMessage.endsWith('?');

    // Optimize isShouting calculation to avoid an extra string allocation from toUpperCase()
    // and allow for early exit when a lowercase letter is found.
    let containsLetter = false;
    let isAllUppercase = true; // Assume true until proven otherwise
    for (let i = 0; i < trimmedMessage.length; i++) {
        const charCode = trimmedMessage.charCodeAt(i);

        // Check for ASCII letters
        const isLowerCaseLetter = (charCode >= 97 && charCode <= 122); // 'a' to 'z'
        const isUpperCaseLetter = (charCode >= 65 && charCode <= 90);  // 'A' to 'Z'

        if (isLowerCaseLetter || isUpperCaseLetter) {
            containsLetter = true;
            if (isLowerCaseLetter) {
                isAllUppercase = false;
                // If a lowercase letter is found, it cannot be shouting (all uppercase),
                // so we can stop processing the string for this condition.
                break;
            }
        }
    }
    const isShouting = containsLetter && isAllUppercase;

    // Apply the logic based on the determined flags
    if (isQuestion && isShouting) {
        return "Calm down, I know what I'm doing!";
    } else if (isQuestion) { // isQuestion is true, isShouting is false
        return 'Sure.';
    } else if (isShouting) { // isQuestion is false, isShouting is true
        return 'Whoa, chill out!';
    } else { // isQuestion is false, isShouting is false
        return 'Whatever.';
    }
}